design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,161 @@
+# Extend VolumePolicies to support more actions
+
+## Abstract
+
+Currently, the [VolumePolicies feature](https://github.com/vmware-tanzu/velero/blob/main/design/Implemented/handle-backup-of-volumes-by-resources-filters.md) which can be used to filter/handle volumes during backup only supports the skip action on matching conditions. Users need more actions to be supported.
+
+## Background
+
+The `VolumePolicies` feature was introduced in Velero 1.11 as a flexible way to handle volumes. The main agenda of introducing the VolumePolicies feature was to improve the overall user experience when performing backup operations for volume resources, the feature enables users to group volumes according the `conditions` (criteria) specified and also lets you specify the `action` that velero needs to take for these grouped volumes during the backup operation. The limitation being that currently `VolumePolicies` only supports `skip` as an action, We want to extend the `action` functionality to support more usable options like `fs-backup-volumes` (File system backup) and `fs-backup-volumes-excludes` (VolumeSnapshots).
+
+## Goals
+- Extending the VolumePolicies to support more actions like file system backup(`fs-backup-volumes`) and snapshots(`fs-backup-volumes-excludes`).
+- Improve user experience when backing up Volumes via Velero
+
+## Non-Goals
+- No changes to existing approaches to opt-in/opt-out annotations for volumes
+- No changes to existing `VolumePolicies` functionalities
+
+
+## Use-cases/Scenarios
+
+**Use-case 1:**
+- A user wants to use `fs-backup-volumes-excludes` (volumesnapshots) backup option for all the csi supported volumes and `fs-backup-volumes` for the rest of the volumes.
+- Currently, velero supports this use-case but the user experience is not that great.
+- The user will have to individually annotate the volume mounting pod with the annotation "backup.velero.io/backup-volumes" for `fs-backup-volumes`
+- This becomes cumbersome at scale.
+- Using `VolumePolicies`, the user can just specify a simple `VolumePolicy` like for csi supported volumes and rest can be defaulted to `fs-backup-volumes`:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    csi:
+      driver: aws.efs.csi.driver
+  action:
+    type: fs-backup-volumes-excludes
+```
+
+**Use-case 2:**
+- A user wants to use `fs-backup-volumes` for nfs volumes
+- In such a scenario the user can just specify a `VolumePolicy` like:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    nfs:
+      server: 192.168.200.90
+  action:
+    type: fs-backup-volumes
+```
+## High-Level Design
```

Comment:
Please also add a paragraph in this section, with some examples, to clarify the priority.  
For example, a PVC may be annotated to `opt-out` but the action is set to `fs-backup-volumes` in the policy.

Reply:
@reasonerjt see my comment at the end -- if we're using policies as another way to implement opt-in/opt-out, then this would work the same way as if a volume had annotations for both. When opt-out is selected (defaultVolumesToFsBackup=true), opt-in annotations are ignored ,and when opt-in is selected (defaultVolumesToFsBackup=false), opt-out annotations are ignored. On the other hand, if we end up deciding *not* to use the opt-in/opt-out workflow for this, then the two approaches are incompatible and we'll have validation to ensure that users only use these new volume policies when defaultVolumesToFsBackup is unset -- sort of like the new includes/excludes fields, users would need to choose to use the new approach or the legacy approach, not both.

Reply:
@shubham-pampattiwar @sseago 
I think this comment remains open.  We need to define the priority of "actions" in this design.  For example, in the spec of the backup CR the `snapshotVolumes` may be set to `true` but the `action` in volumepolicy is set to `fs-backup`, or vice versa.  Things can be more complicated if we take the `opt-in/out` annotations into account.

Reply:
@reasonerjt From what we agreed to on the call, policies take precedence here -- we need to be explicit in the design. For any volume that has a matching volume policy, the opt-in/out annotations and defaults will not be used. If a "skip" policy matches, the volume is skipped, if a "snapshot" policy matches, we use CSI or native snapshots, if a "fs-backup" policy matches, we use fs-backup. If none match, then we follow the opt-in/out logic based on annotations and defaultVolumesToFsBackup param.

Reply:
As for priority among policy actions, the first matched policy should take priority.  See the above example policies -- the more specific example is first -- "gp2" storageclass volumes have "snapshot" policy, and then a catch-all "fs-backup" policy that matches all volumes. This means all gp2 volumes will be "snapshot", since that's the first match, and all non-gp2 volumes will be "fs-backup". Since these policies match *all* volumes, we don't have to resort to the fallback at all. If we didn't have that second policy, then all gp2 volumes would be "snapshot", and all other volumes would use the fallback of the legacy opt-in/out approach.

Reply:
Agreed.
@shubham-pampattiwar could you please update the design to explicitly explain the priority?